README
README for JTE version @VERSION@

Steve McIntyre <steve@einval.com>
@DATE@

License - GPL v2+. See the file COPYING for more details.

JTE - Jigdo Template Export
===========================

  • Introduction - jigdo and JTE
  • Download
  • How to use JTE
  • How JTE works
  • How to use jigit-mkimage
  • (Dead) experiments
  • External integration
  • What's left to do?

----------------------------------------------------------------------

Introduction - jigdo and JTE
----------------------------

Jigdo is a useful tool to help in the distribution of large files like CD
and DVD images. See Richard Atterer's site [1] for more details. Debian CDs
and DVD ISO images are published on the web in jigdo format to allow end
users to download them more efficiently.

[1] http://atterer.org/jigdo/

Jigdo is generic and powerful - it can be used for any large files that
are made up of smaller files. However, to be this generic is costly.
Creating jigdo files from ISO images is quite inefficient - to work out
which files are included in the ISO image, jigdo has to calculate and
compare checksums of every possible file and every extent in the image.
Essentially it has to brute-force the image. It can take a long time to do
this for a large image (imagine a 4.5GB DVD image or a 30+GB Blu-Ray
image).

I first started looking for ways to improve this back in 2004:

 1. Modify jigdo so it knew about the internals of ISO images and could
    efficiently scan them (bad, not very generic for jigdo)
 2. Write a helper tool to dump extra information for jigdo to use
    alongside the ISO image (I had a helper tool written, but modifying
    jigdo to use this looked HARD)
 3. Patch mkisofs/genisoimage to write .jigdo and .template files
    alongside the ISO image

I completed the third of these options, and called it JTE (or Jigdo
Template Export). The code worked fine, and ran in a very small fraction
of the time taken to run genisoimage and jigdo separately. The output
.jigdo and .template files worked correctly, i.e. jigdo-file and the
wrapper script jigdo-mirror accept them and would generate an ISO image
that exactly matches the original.

Debian used that code for a number of years within genisoimage, but we've
since switched over to using xorriso [2] instead for our image building
instead. It has a lot of useful features that we want compared to
genisoimage, not least a friendly and engaged author in Thomas Scmitt!

[2] https://www.gnu.org/software/xorriso/

Thomas and I and George Danchev worked together to package up my JTE code
into libjte such that xorriso could use it effectively. Xorriso has been
capable of generating jigdo files since 2010.

In late 2019, I took over maintenance of the jigdo upstream code and added
support for a new (v2) jigdo data format, using SHA256 instead of MD5
internally. See my jigdo page for more details about that. I have also
updated the JTE codebase to support this new format, of course.

As genisoimage is effectively dead at this point, I took the decision to
not add the jigdo v2 support into the genisoimage codebase. If you need to
generate jigdo v2 format, either use jigdo itself or xorriso if you'd like
the performance benefit of the libjte integration.

JTE includes a few tools:

  • jigit-mkimage, a simple and very fast tool to reconstruct image files
    from .jigdo and .template files. It doesn't have any logic to cope
    with downloading missing files, but will list the missing files that
    are needed. It is also much faster for people (like me!) who already
    have full local mirrors.
  • parallel-sums is a simple extra utility to generate checksums quickly
    and efficiently, reading file data only once and calculating checksums
    using multiple algorithms in parallel using threads.
  • jigsum, jigsum-sha256 and rsyncsum are checksum tools which will
    output checksums in jigdo's base64-like format rather than the normal
    hexadecimal format. Useful for debugging jigdo issues.
  • jigdump is a tool to dump the contents of a jigdo template or .iso.tmp
    file. Useful for debugging jigdo issues.
  • mkjigsnap is a utility to help with maintaining the "snapshots" that
    jigdo needs if you're going to be keeping data around for users in the
    long term. We use this on some Debian systems.

Why the "jigit" name? The packages and source are named jigit to match the
name of a long-dead wrapper script. That script may be gone, but it's
easier to keep the name!

----------------------------------------------------------------------

Download
--------

The jigit source package (and hence the various binary packages it builds) is
included in the main Debian archive, so your best bet is to get binary packages
from there. Check for the current version(s) using tracker.debian.org).

Source and backported versions are in the download area [3] alongside
the current ChangeLog. All the files for download are PGP-signed for
safety. You can find my keys online if you need them [4].

jigit is maintained in git [5].

[3] https://www.einval.com/~steve/software/JTE/download/
[4] https://www.einval.com/~steve/pgp/
[5] https://git.einval.com/cgi-bin/gitweb.cgi?p=jigit.git.

----------------------------------------------------------------------

How to use JTE
--------------

To use the jigdo creation code in xorriso, add some extra command line
options to control the jigdo features. You must specify the location of
the output .jigdo and .template files alongside the ISO image, and a
"checksum" list file, containing the checksums that you want JTE to match.
You can also specify a lot of other options to control the contents of the
.jigdo file. A complicated (but realistic) example from my own test setup
is here, with all the extra jigdo parameters explained below:

xorriso -as mkisofs -r -J \
        -V 'Debian TEST amd64 n' \
        -o debian-TEST-amd64-NETINST-1.iso \
        -jigdo-jigdo debian-TEST-amd64-NETINST-1.jigdo \
        -jigdo-template debian-TEST-amd64-NETINST-1.template \
        -checksum_algorithm_iso sha256,sha512 \
        -checksum-list /tmp/buster/checksum-check \
        -jigdo-checksum-algorithm md5 \
        -jigdo-force-checksum /pool/ \
        -jigdo-min-file-size 1024 \
        -jigdo-exclude 'README*' \
        -jigdo-exclude /doc/ \
        -jigdo-exclude /md5sum.txt \
        -jigdo-exclude /.disk/ \
        -jigdo-exclude /pics/ \
        -jigdo-exclude 'Release*' \
        -jigdo-exclude 'Packages*' \
        -jigdo-exclude 'Sources*' \
        -jigdo-exclude boot1 \
        -jigdo-map Debian=/scratch/mirror/debian/ \
        -joliet-long -cache-inodes \
        -isohybrid-mbr syslinux/usr/lib/ISOLINUX/isohdpfx.bin \
        -b isolinux/isolinux.bin \
        -c isolinux/boot.cat \
        -boot-load-size 4 \
        -boot-info-table -no-emul-boot \
        -eltorito-alt-boot -e boot/grub/efi.img \
        -no-emul-boot -isohybrid-gpt-basdat \
        -isohybrid-apm-hfsplus boot1 CD1

That's a long command line, but it's not too hard to follow:

  • -jigdo-jigdo specifies the output filename for the .jigdo file
  • -jigdo-template specifies the output filename for the .template file
  • -checksum_algorithm_iso specifies which checksums to include for the
    ISO image inside the .jigdo file
  • -checksum-list specifies the input filename for the checksum data that
    JTE needs
  • -jigdo-checksum-algorithm specifies which checksum algorithm to use
    inside the jigdo file, both for describing the files in the ISO and
    the ISO itself. The allowed options are "md5" (i.e. jigdo format v1,
    the default), or "sha256" (i.e. jigdo format v2).
  • -jigdo-force-checksumadds a match pattern for the jigdo generation
    code - all files matching this pattern must be listed in the
    checksum-list, and they mush have the correct checksum. This is used
    in Debian as a precaution that the source data is correct for all the
    packages we're including on our media.
  • -jigdo-min-file-size and -jigdo-exclude are two different ways to stop
    certain files from being matched in the jigdo generation code. We
    don't want to waste time on files that are too small, or too temporary
    (e.g. generated during the CD build process itself), or that are not
    tracked cleanly with versions inside the Debian archive. Files
    excluded from the jigdo generation using these parameters will
    therefore be included directly in the .template raw data section.
  • Finally, one or more -jigdo-map entries should be added, to map
    pathnames in the .jigdo file to the [Servers] section.

If the -jigdo-* options are not used, the normal xorriso execution path is
not affected at all. The above invocation will create 3 output files
(.iso, .jigdo and .template). Multiple -jigdo-exclude and -jigdo-map
options are accepted, for multiple exclude and map patterns.

----------------------------------------------------------------------

How JTE works
-------------

Internally in libisoburn (and hence xorriso), in all the places where it
will write image data it will also call into libjte to offer that image
data for jigdo processing. Any file data entries are passed through with
information about the original file. If that file is not excluded (because
of its path or size, as mentioned), JTE will grab the filename, the size
of the file and the checksum of the file's data. If that checksum, size
and length match an entry in the input checksum-list, JTE will write a
file match record into the template file (and then the jigdo file) instead
of the file data itself. For anything else (excluded files, directory
data, etc.), raw data is simply copied through and compressed into the
template file.

----------------------------------------------------------------------

How to use jigit-mkimage
------------------------

jigit-mkimage is a faster, more minimal version of "jigdo-file
make-image", written in portable C. It takes a few options:

┌─────────┬──────────────────────────────────────────────────────────────┐
│-f <MD5  │Specify a file containing MD5sums for files we should attempt │
│file>    │to use when rebuilding the image                              │
├─────────┼──────────────────────────────────────────────────────────────┤
│-F       │Specify a file containing SHA256sums for files we should      │
│<SHA256  │attempt to use when rebuilding the image                      │
│file>    │                                                              │
├─────────┼──────────────────────────────────────────────────────────────┤
│-j <jigdo│Specify the input jigdo file                                  │
│file>    │                                                              │
├─────────┼──────────────────────────────────────────────────────────────┤
│-t       │                                                              │
│<template│Specify the input template file                               │
│file>    │                                                              │
├─────────┼──────────────────────────────────────────────────────────────┤
│-m <item=│Map <item> to <path> to find the files in the mirror          │
│path>    │                                                              │
├─────────┼──────────────────────────────────────────────────────────────┤
│-M       │Don't attempt to build the image; just verify that all the    │
│<Missing │components needed are available. If some are missing, list    │
│file>    │them in the specified file.                                   │
├─────────┼──────────────────────────────────────────────────────────────┤
│-v       │Make the output logging more verbose.                         │
├─────────┼──────────────────────────────────────────────────────────────┤
│-l <log  │Specify a logfile. If not specified, will log to stderr just  │
│file>    │like genisoimage                                              │
├─────────┼──────────────────────────────────────────────────────────────┤
│         │Don't bother checking checksums of the input files, or of the │
│-q       │output image.                                                 │
│         │WARNING: this may lead to corrupt images, but is faster as    │
│         │less work is done.                                            │
├─────────┼──────────────────────────────────────────────────────────────┤
│-s <start│Specify where to start in the image (in bytes). If not        │
│offset>  │specified, will start at the beginning (offset 0). Added for  │
│         │iso-image.pl use                                              │
├─────────┼──────────────────────────────────────────────────────────────┤
│-e <end  │Specify where to end in the image (in bytes). If not          │
│offset>  │specified, will run all the way to the end of the image. Added│
│         │for iso-image.pl use                                          │
├─────────┼──────────────────────────────────────────────────────────────┤
│         │Don't attempt to reassemble the image; simply parse the image │
│-z       │descriptor in the template file and print the image size.     │
│         │Added for iso-image.pl use                                    │
└─────────┴──────────────────────────────────────────────────────────────┘

Specifying a start or end offset implies -q - it's difficult to check
checksums if the full image is not generated!

----------------------------------------------------------------------

(Dead) experiments
------------------

I had extra plans for JTE that never really came to fruition due to a lack
of time and energy... :-/ Check git history if you're interested.

iso-image.pl - on-the-fly rebuild of ISO images for HTTP

iso-image.pl was a small perl wrapper script written to drive mkimage and
turn it into a CGI. It would parse the incoming request (including
byte-ranges) and call jigit-mkimage to actually generate the image pieces
wanted.

This code worked, but was always too slow for production use. Each CGI
request needed to index into the ISO image independently, leading to lots
and lots of overlapping calls to decompress the template data.

jigdoofus - a better way to do on-the-fly assembly

I started on a new project, creating a FUSE-based filesystem that would
rebuild ISOs on the fly. I decided to use a database backend and a caching
system to solve the problem of the repetitive decompression that stopped
iso-image.pl. I made some progress, but ran out of steam. Code is still in
the "jigdoofus" branch in git in case anybody ever finds it useful.

jigit - a friendly wrapper for jigit-mkimage

Similarly to the jigdo-lite script in the jigdo package, I wanted to
provide a nicer user experience for easy downloading of Debian and Ubuntu
CD images. It worked, but never really gained much traction. It needed
much more effort to make things reliable for production use.

----------------------------------------------------------------------

External integration
--------------------

debian-cd

The debian-cd package in Debian is what we use to generate installer CDs
and DVDs. It has supported JTE since 2005, and we still use it every day.

cdrkit/genisoimage

genisoimage in Debian shipped with integrated JTE code for a long time,
but is basically dead upstream. Not recommended for use any more.

xorriso

xorriso uses libjte to generate jigdo and template files, and has worked
this way since 2010.

----------------------------------------------------------------------

What's left to do?
------------------

 1. Testing! :-) This is where you lot come in! Please play with this some
    more and let me know if you have any problems, especially with data
    corruption.
 2. More documentation.


ChangeLog
Version 1.22 (2019-11-21)

 * Add support for SHA256 checksums, using jigdo format v2
   + Changes to libjte to generate the new format as an option - still
     defaults to v1 for cmpatibility for now
   + libjte updated to major version 2
   + Add support for v2 in jigdump, jigit-mkimage etc. too
   + Add new jigsum-sha256 program, for the base64-like output with
     sha256 checksums
 * jigdump now also understands jigdo .iso.tmp files
 * Misc small cleanups

Version 1.21 (2019-02-04)

 * parallel-sums:
   + New utility for calculating checksums in parallel
 * Misc code cleanups
 * Misc man page cleanups
 * jigit-mkimage:
   + Print missing file names correctly

Version 1.20 (2014-08-18)

 * mkjigsnap:
   + Add support for tracking back-references, i.e. which jigdo did a
     particular file come from.
   * Add support for verifying checksums
 * jigsum:
   * Add support for verifying checksums, similar to md5sum etc.
 * Misc code cleanups:
   * warnings
   * sparse errors
   * Fix memset calls to zero the right amount of memory

Version 1.19 (2011-06-16)

 * Improved error-handling in mkimage and libjte.
 * mkjigsnap rewrite:
   + merge two old implementations
   + switch to perl: cleaner code, much faster snapshot generation
   + added -v (verbose) and -N (dry-run) options
 * Put AC_CONFIG_MACRO_DIR in a better place in the libjte configure
   script.
 * Remove -Werror from the Makefile - gcc is getting more and more
   pedantic with every release, and this may cause package builds to fail.

Version 1.18 (2011-04-30)

 * More dynamic memory handling in mkimage, instead of allocating
   fixed-size strings on the stack. Should now work on Hurd too.
 * Fix man page syntax errors.
 * s,config.h,../config.h, in libjte for easier use by the xorriso folks

Version 1.17 (2011-03-16)

 * Rename mkimage to jigit-mkimage, to avoid a naming clash.
 * Sanity-check jigdo files - require a valid JigsawDownload
   header by default.
 * Multiple fixes for how MD5 comparisons are dealt with.
 * Add libjte - jigdo generation code that used to be embedded in
   mkisofs/genisoimage. Now a shared library useable in xorriso etc.

Version 1.16 (2007-10-18)

 * Remove build-stamp from the release tarball, to make debian package
   builds work better
 * Fix wget options typo
 * Add proper working bzip2-compressed template support in jigdump and
   mkimage
 * In mkimage, don't stat every file mentioned in the md5 list until we
   need to.
 * Other small cleanups in jigdump and mkimage

Version 1.15 (2005-01-17)

 * Minor compilation fixes
 * Cope with missing files better
 * Minor jigdo file-handling cleanups
 * mkimage code refactored to make it easier to maintain
 * Cleaned up licensing - added explicit GPL v2 headers all over
 * Fix image reconstruction with jigit

Version 1.14 (2004-10-12)

 * Fixed generation of rsyncsums
 * Added -d option to mkjigsnap to specify DIRNAME
 * Minor prettiness tweaks to output UI of mkimage and jigit
 * 64-bit printf fixes

Version 1.13 (2004-09-05)

 * Improved the documentation, listing new options to mkimage
 * Fixed/updated the debian-cd patch

Version 1.12 (2004-09-05)

 * Improved the documentation, listing new options to mkimage
 * Fixed/updated the debian-cd patch

Version 1.11 (2004-09-05)

 * Added rsync sums into JTE output
 * Added jigit script
 * Added extra options to mkimage to support jigit

Version 1.10 (2004-09-01)

 * Removed the last md5 check; no longer needed
 * Added new -jigdo-force-md5 option; any files matching specified
   paths MUST match an MD5 supplied, or fail
 * Fixed a couple of typos.
 * Added initial work for bz2 support.
 * debian-cd diff now includes updated grab_md5 and jigdo_cleanup
   scripts that must be used.
 * mkisofs now leaves a blank [Servers] section in the .jigdo file -
   jigdo_cleanup does the work.

Version 1.9 (2004-08-30)

 * Added MD5 reverse lookups to pick up boot files; helps reduce
   CD/DVD #1 size substantially, BUT -md5-list option now mandatory.
 * Fixed sizing bug in HFS hybrid template code.

Version 1.8 (2004-08-27)

 * Tweaked the -exclude option to reduce template file sizes.

Version 1.7 (2004-08-19)

 * Fixed a couple of silly bugs introduced; don't assume template
   creation.
 * Added README files and man page sections for the new boot
   support.

Version 1.6 (2004-07-14)

 * Added code to check MD5 sums in mkisofs; should no longer need
   to run mirror_check any more. See the -md5-list option.

Version 1.5 (2004-07-05)

 * Minor bug-fixes and updates to mkimage
 * Added iso-image.pl wrapper script

Version 1.4 (2004-06-26)

 * Added mkimage
 * Moved utils and patch into a tarball

Version 1.3 (2004-06-18)

 * Fixed crash without jigdo options
 * Cleaned up mipsel boot code
 * Make it easier to specify the boot command line for hppa
 * Minor cosmetic .jigdo cleanup

Version 1.2 (2004-06-14)

 * Added boot support for alpha, hppa, mips, mipsel
 * Cosmetic changes to the .jigdo file

Version 1.1 (2004-06-10)

 * Removed .jte file code
 * Large cleanup to reduce patch size

Version 1.0 (2004-06-07)

 * Initial release
 * Produced .jigdo, .template and .jte files


jigdump.c
/*
 * jigdump
 *
 * Tool to dump the contents of a jigdo template file
 *
 * Copyright (c) 2004-2019 Steve McIntyre <steve@einval.com>
 *
 * GPL v2 - see COPYING
 */

#include <errno.h>
#include <math.h>
#include <stdlib.h>
#include <string.h>
#include <zlib.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/mman.h>
#include "jig-base64.h"
#include "md5.h"
#include "jigdo.h"

#define HEADER_STRING "JigsawDownload template"

static INT64 find_string(unsigned char *buf, size_t buf_size, char *search)
{
    size_t length = strlen(search);
    INT64 result;
    
    for (result = 0; result < (buf_size - length); result++)
    {
        if (!memcmp(&buf[result], search, length))
            return result;
    }
    return -1;
}

static INT64 parse_data_block(INT64 offset, unsigned char *buf, size_t buf_size)
{
    /* Parse the contents of this data block... */
    UINT64 dataLen = 0;
    UINT64 dataUnc = 0;

    if (!strncmp((char *)buf, "DATA", 4))
        printf("\ngzip data block found at offset %lld\n", offset);
    else
        printf("\nbzip2 data block found at offset %lld\n", offset);

    dataLen = (UINT64)buf[4];
    dataLen |= (UINT64)buf[5] << 8;
    dataLen |= (UINT64)buf[6] << 16;
    dataLen |= (UINT64)buf[7] << 24;
    dataLen |= (UINT64)buf[8] << 32;
    dataLen |= (UINT64)buf[9] << 40;
    printf("  compressed block size %llu bytes\n", dataLen);

    dataUnc = (UINT64)buf[10];
    dataUnc |= (UINT64)buf[11] << 8;
    dataUnc |= (UINT64)buf[12] << 16;
    dataUnc |= (UINT64)buf[13] << 24;
    dataUnc |= (UINT64)buf[14] << 32;
    dataUnc |= (UINT64)buf[15] << 40;
    printf("  uncompressed block size %llu bytes\n", dataUnc);

    return dataLen;
}

static INT64 parse_desc_block(INT64 offset, unsigned char *buf, size_t buf_size)
{
    /* Parse the contents of this data block... */
    UINT64 descLen = 0;
    
    printf("\nDESC block found at offset %lld (%llx)\n", offset, offset);
    descLen = (UINT64)buf[4];
    descLen |= (UINT64)buf[5] << 8;
    descLen |= (UINT64)buf[6] << 16;
    descLen |= (UINT64)buf[7] << 24;
    descLen |= (UINT64)buf[8] << 32;
    descLen |= (UINT64)buf[9] << 40;
    printf("  DESC block size is %llu bytes\n", descLen);
    
    return 10;
}

static INT64 parse_desc_data(INT64 offset, unsigned char *buf, size_t buf_size)
{
    int type = buf[0];
    printf("  DESC entry: block type %d at offset %lld (0x%llx)\n",
	   type, offset, offset);
    
    switch (type)
    {
        case BLOCK_DATA:
        {
            UINT64 skipLen = 0;
            skipLen = (UINT64)buf[1];
            skipLen |= (UINT64)buf[2] << 8;
            skipLen |= (UINT64)buf[3] << 16;
            skipLen |= (UINT64)buf[4] << 24;
            skipLen |= (UINT64)buf[5] << 32;
            skipLen |= (UINT64)buf[6] << 40;
            printf("    7 byte entry:\n");
            printf("      Template data, %llu bytes of data\n", skipLen);
            return 7;
        }
        case BLOCK_IMAGE_MD5:
        {
            UINT64 imglen = 0;
            UINT32 blocklen = 0;
            int i = 0;

            imglen = (UINT64)buf[1];
            imglen |= (UINT64)buf[2] << 8;
            imglen |= (UINT64)buf[3] << 16;
            imglen |= (UINT64)buf[4] << 24;
            imglen |= (UINT64)buf[5] << 32;
            imglen |= (UINT64)buf[6] << 40;

            blocklen = (UINT32)buf[23];
            blocklen |= (UINT32)buf[24] << 8;
            blocklen |= (UINT32)buf[25] << 16;
            blocklen |= (UINT32)buf[26] << 24;

	    printf("    27 byte entry:\n");
	    printf("      Image info block v1 using MD5\n");
            printf("      Original image length %llu bytes\n", imglen);
            printf("      Image MD5: ");
            for (i = 7; i < 23; i++)
                printf("%2.2x", buf[i]);
            printf(" (%s)\n", base64_dump(&buf[7], 16));
            printf("      Rsync block length %lu bytes\n", blocklen);
            return 27;
        }
        case BLOCK_IMAGE_SHA256:
        {
            UINT64 imglen = 0;
            UINT32 blocklen = 0;
            int i = 0;

            imglen = (UINT64)buf[1];
            imglen |= (UINT64)buf[2] << 8;
            imglen |= (UINT64)buf[3] << 16;
            imglen |= (UINT64)buf[4] << 24;
            imglen |= (UINT64)buf[5] << 32;
            imglen |= (UINT64)buf[6] << 40;

            blocklen = (UINT32)buf[39];
            blocklen |= (UINT32)buf[40] << 8;
            blocklen |= (UINT32)buf[41] << 16;
            blocklen |= (UINT32)buf[42] << 24;

	    printf("    43 byte entry:\n");
	    printf("      Image info block v2 using SHA256\n");
            printf("      Original image length %llu bytes\n", imglen);
            printf("      Image SHA256: ");
            for (i = 7; i < 39; i++)
                printf("%2.2x", buf[i]);
            printf(" (%s)\n", base64_dump(&buf[7], 32));
            printf("      Rsync block length %lu bytes\n", blocklen);
            return 43;
        }
        case BLOCK_MATCH_MD5:
        case BLOCK_WRITTEN_MD5:
        {
            UINT64 fileLen = 0;
            int i = 0;

            fileLen = (UINT64)buf[1];
            fileLen |= (UINT64)buf[2] << 8;
            fileLen |= (UINT64)buf[3] << 16;
            fileLen |= (UINT64)buf[4] << 24;
            fileLen |= (UINT64)buf[5] << 32;
            fileLen |= (UINT64)buf[6] << 40;
            
	    printf("    31 byte entry:\n");
            printf("      File %s block v1 using MD5\n",
                   (type == BLOCK_MATCH_MD5 ? "match" : "written"));
            printf("      length %llu bytes\n", fileLen);
            printf("      file rsyncsum: ");
            for (i = 7; i < 15; i++)
                printf("%2.2x", buf[i]);
            printf(" (%s)\n", base64_dump(&buf[7], 8));
            printf("      file md5: ");
            for (i = 15; i < 31; i++)
                printf("%2.2x", buf[i]);
            printf(" (%s)\n", base64_dump(&buf[15], 16));
            return 31;
        }
        case BLOCK_MATCH_SHA256:
        case BLOCK_WRITTEN_SHA256:
        {
            UINT64 fileLen = 0;
            int i = 0;

            fileLen = (UINT64)buf[1];
            fileLen |= (UINT64)buf[2] << 8;
            fileLen |= (UINT64)buf[3] << 16;
            fileLen |= (UINT64)buf[4] << 24;
            fileLen |= (UINT64)buf[5] << 32;
            fileLen |= (UINT64)buf[6] << 40;

	    printf("    47 byte entry:\n");
            printf("      File %s block v2 using SHA256\n",
                   (type == BLOCK_MATCH_SHA256 ? "match" : "written"));
            printf("      length %llu bytes\n", fileLen);
            printf("      file rsyncsum: ");
            for (i = 7; i < 15; i++)
                printf("%2.2x", buf[i]);
            printf(" (%s)\n", base64_dump(&buf[7], 8));
            printf("      file sha256: ");
            for (i = 15; i < 47; i++)
                printf("%2.2x", buf[i]);
            printf(" (%s)\n", base64_dump(&buf[15], 32));
            return 47;
        }
        default:
	    printf("    Unrecognised block type %d, skipping forwards in hope\n", type);
	    return 1;
            break;
    }

    return 0;
}

static INT64 parse_desc_pointer(INT64 offset, unsigned char *buf, size_t buf_size, INT64 template_size, INT64 desc_offset)
{
    /* Parse the contents of this final block... */
    UINT64 ptr = 0;

    ptr = (UINT64)buf[0];
    ptr |= (UINT64)buf[1] << 8;
    ptr |= (UINT64)buf[2] << 16;
    ptr |= (UINT64)buf[3] << 24;
    ptr |= (UINT64)buf[4] << 32;
    ptr |= (UINT64)buf[5] << 48;
    printf("  DESC pointer offset %llu\n", ptr);
    printf("  points to DESC at offset %llu\n", template_size - ptr);
    printf("  actual DESC section starts at %llu\n", desc_offset);
    if (template_size - ptr != desc_offset)
	printf("ERROR: pointers don't match up\n");

    return 6;
}

static INT64 parse_tmp_desc_pointer(unsigned char *buf)
{
    /* Parse the contents of this final block... */
    UINT64 ptr = 0;

    ptr = (UINT64)buf[0];
    ptr |= (UINT64)buf[1] << 8;
    ptr |= (UINT64)buf[2] << 16;
    ptr |= (UINT64)buf[3] << 24;
    ptr |= (UINT64)buf[4] << 32;
    ptr |= (UINT64)buf[5] << 48;

    return ptr;
}

static INT64 parse_header_block(INT64 offset, unsigned char *buf, size_t buf_size)
{
    /* Massive overkill for sizes, but safe! */
    char format_version[BUF_SIZE] = {0};
    char generator[BUF_SIZE] = {0};
    int ret = 0;

    ret = sscanf((char *)buf, HEADER_STRING" %s %s", format_version, generator);
    if (ret != 2)
    {
        printf("Failed to parse header, ret %d\n", ret);
	return -1;
    }
    printf("Found header: \"%s\"\n", HEADER_STRING);
    printf("Format version: %s\n", format_version);
    printf("Generator: %s\n", generator);

    return 0;
}

/* return 1 if haystack ends with the text contained in needle, 0
 * otherwise */
static int string_ends(char *haystack, char *needle)
{
    int hlen;
    int nlen;

    if (NULL == haystack || NULL == needle)
        return 0;

    hlen = strlen(haystack);
    nlen = strlen(needle);
    if (nlen > hlen)
        return 0;
    if (!strcmp((haystack + hlen - nlen), needle))
        return 1;

    return 0;
}

int main(int argc, char **argv)
{
    char *filename = NULL;
    int fd = -1;
    unsigned char *buf = NULL;
    INT64 offset = 0;
    INT64 bytes = 0;
    INT64 template_size = 0;
    INT64 desc_offset = 0;
    struct stat sb;
    e_state state = STARTING;
    
    if (argc != 2)
    {
        printf("No filename specified! Try again...\n");
        return EINVAL;
    }
    
    filename = argv[1];
    
    fd = open(filename, O_RDONLY);
    if (-1 == fd)
    {
        printf("Failed to open file %s, error %d!. Try again...\n", filename, errno);
        return errno;
    }

    if (-1 == fstat(fd, &sb))
    {
        printf("Failed to stat file %s, error %d!. Try again...\n", filename, errno);
        return errno;
    }
    template_size = sb.st_size;

    buf = malloc(BUF_SIZE);
    if (!buf)
    {
        printf("Failed to malloc %d bytes. Abort!\n", BUF_SIZE);
        return ENOMEM;
    }

    /* If we're looking at a template file, find the beginning of the
     * data - read the first chunk, including the header. Only do this
     * if the filename ends in ".template" */
    if (string_ends(filename, ".template"))
    {
        printf("Filename %s ends in .template.\n", filename);
        printf("Assuming this is meant to be a template file.\n");
	while (STARTING == state)
	{
	    INT64 start_offset = -1;

	    bytes = read(fd, buf, BUF_SIZE);
	    if (0 >= bytes)
	    {
		state = DONE;
		break;
	    }

	    start_offset = find_string(buf, bytes, HEADER_STRING);
	    if (0 != start_offset)
	    {
		printf("Can't find header - this is not a template file\n");
		return EINVAL;
	    }
	    if (0 != parse_header_block(start_offset, buf, bytes))
		return EINVAL;

	    start_offset = find_string(buf, bytes, "DATA");
	    if (-1 == start_offset)
		start_offset = find_string(buf, bytes, "BZIP");
	    if (start_offset >= 0)
	    {
		offset += start_offset;
		state = IN_DATA;
		break;
	    }
	    offset += bytes;
	}
    }
    else
    {
        printf("Filename %s does not end in .template.\n", filename);
        printf("Assuming this is NOT meant to be a template file.\n");
        /* We're looking at a tmp file? No header available. Seek to
        the end for the final DESC * pointer */
        lseek(fd, template_size - 6, SEEK_SET);
        bytes = read(fd, buf, BUF_SIZE);
        if (bytes != 6)
        {
            state = DONE;
        }
        else
        {
            INT64 start_offset = parse_tmp_desc_pointer(buf);
            offset = template_size - start_offset;
            state = IN_DESC;
        }
    }

    while (DONE != state && ERROR != state)
    {
        INT64 start_offset = -1;
        lseek(fd, offset, SEEK_SET);
        bytes = read(fd, buf, BUF_SIZE);
        if (0 >= bytes)
        {
            state = ERROR;
            break;
        }
        if (IN_DATA == state)
        {
            if (!find_string(buf, bytes, "DATA") || !find_string(buf, bytes, "BZIP"))
                state = IN_DATA;
            if (!find_string(buf, bytes, "DESC"))
	    {
                state = IN_DESC;
		desc_offset = offset;
	    }
        }
        switch (state)
        {
            case IN_DATA:
                start_offset = parse_data_block(offset, buf, bytes);
                offset += start_offset;
                break;
            case IN_DESC:
                start_offset = parse_desc_block(offset, buf, bytes);
                offset += start_offset;
                state = DUMP_DESC;
                break;
            case DUMP_DESC:
                start_offset = parse_desc_data(offset, buf, bytes);
                offset += start_offset;
		if (offset >= (template_size - 6))
			state = DUMP_DESC_PTR;;
                break;
	    case DUMP_DESC_PTR:
		state = DONE;
		start_offset = parse_desc_pointer(offset, buf, bytes, template_size, desc_offset);
		offset += start_offset;
		break;
            default:
                break;
        }
	if (offset == template_size)
	    state = DONE;
    }        
    
    close(fd);

    return 0;
}
